pkg/apis/core/validation/validation.go

```diff
@@ -4294,6 +4302,73 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 	return allErrs
 }
 
+func validatePodResources(spec *core.PodSpec, podClaimNames sets.Set[string], fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+	if spec.Resources == nil {
+		return nil
+	}
+
+	allErrs := field.ErrorList{}
+
+	// ValidateResourceRequirements checks if resource names and quantities are
+	// valid, and requests are less than limits.
+	allErrs = append(allErrs, ValidatePodResourceRequirements(spec.Resources, podClaimNames, fldPath, opts)...)
+	allErrs = append(allErrs, validatePodResourceConsistency(spec, fldPath)...)
+	return allErrs
+}
+
+// validatePodResourceConsistency checks if aggregate container-level requests are
+// less than or equal to pod-level requests, and individual container-level limits
+// are less than or equal to pod-level limits.
+func validatePodResourceConsistency(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Convert the *core.PodSpec to *v1.PodSpec to satisfy the call to
+	// resourcehelper.PodRequests method, in the subsequent lines,
+	// which requires a *v1.Pod object (containing a *v1.PodSpec).
+	v1PodSpec := &v1.PodSpec{}
+	// TODO(ndixita): Convert_core_PodSpec_To_v1_PodSpec is risky. Add a copy of
+	// AggregateContainerRequests against internal core.Pod type for beta release of
+	// PodLevelResources feature.
+	if err := corev1.Convert_core_PodSpec_To_v1_PodSpec(spec, v1PodSpec, nil); err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath, fmt.Errorf("invalid %q: %v", fldPath, err.Error())))
+	}
+
+	reqPath := fldPath.Child("requests")
+	// resourcehelper.AggregateContainerRequests method requires a Pod object to
+	// calculate the total requests requirements of a pod. Hence a Pod object using
+	// v1PodSpec i.e. (&v1.Pod{Spec: *v1PodSpec}, is created on the fly, and passed
+	// to the AggregateContainerRequests method to facilitate proper resource
+	// calculation without modifying AggregateContainerRequests method.
+	aggrContainerReqs := resourcehelper.AggregateContainerRequests(&v1.Pod{Spec: *v1PodSpec}, resourcehelper.PodResourcesOptions{})
+
+	// Pod-level requests >= aggregate requests of all containers in a pod.
+	for resourceName, ctrReqs := range aggrContainerReqs {
```

Comment:
You say that we are limiting this feature to cpu and memory but I don't really see any kind of enforcement in this code.

Should we block folks from setting anything other than cpu/memory for pod resources?

Reply:
PTAL: https://github.com/kubernetes/kubernetes/pull/128407#discussion_r1833133704

Reply:
That sounds good. It may be worth having a few unit tests for validation that test what error folks get if they have an unsupported value.

Reply:
I have the unit tests in place. Although they validation that the unsupported resources throw an error. They don't check the exact error message. Was trying to cater to how the existing tests are written :/

I will send a follow up PR for cleanup so that these tests are easier to identify. Marking the comment as resolved. Feel free to re-open it.